oca/api.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+# Copyright 2004-2015 Odoo S.A.
+# Copyright 2016-2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from decorator import decorator
+
+
+__all__ = [
+    'aggregate', 'foreach',
+]
+
+
+def aggregate(method, value, self):
+    """ Aggregate record-style `value` for a method decorated w/ `@foreach`.
+    """
+    spec = getattr(method, '_returns', None)
+    if spec:
+        # value is a list of instances, concatenate them
+        model, _, _ = spec
+        if model == 'self':
+            return sum(value, self.browse())
+        elif model:
+            return sum(value, self.env[model])
+    return value
+
+
+def foreach(method):
+    """ Loop the decorated method so ``self`` is singleton, return results.
+
+    The values returned from the method are concatenated into a :type:`list`
+    and subsequently returned.
+
+    Example:
+
+        .. code-block:: python
+
+            @api.foreach
+            def method(self):
+                return self.id
+
+        If the above method is called on a Record Singleton with the ``id`` 1
+        it would return:
+
+            [1]
+
+        If the method is called on a RecordSet containing Records with ``id``
+        1, 2, and 3 it would return:
+
+            [1, 2, 3]
```

Comment:
I tried to do this, but was butting heads against the way these decorators work.

In the instance of the record decorators like `multi` & `foreach`, the decorator is being called when the method is being called, and the method is being passed as an implicit parameter.

In order to allow for this feature, the decorator has to be called before the method is called, and thus I cannot receive the method as an implicit parameter.

I guess technically it could be a lambda, but that seems pretty dirty IMO. Pseudocode looks like this, although I doubt it actually works:

```
@lambda m: oca.foreach(m, list)
def my_meth(self):
    return self.id
```

Due to this, I think that this feature would be better as an addendum to the existing `api.returns` instead. We would then use them in parallel:

```
@oca.foreach
@oca.returns(list)
def my_meth(self):
```

Unless you know something about decorators that I don't?